go-controller/pkg/node/vrfmanager/vrf_manager.go

```diff
@@ -0,0 +1,309 @@
+					return err
+				}
+				select {
+				case linkUpdateEvent, ok := <-linkUpdateCh:
+					unsubscribe(doneCh, linkUpdateCh)
+					linkSyncTimer.Reset(60 * time.Second)
+					if !ok {
+						klog.Errorf("Vrf manager: failed to receive interface update event")
+						return nil
+					}
+					ifName := linkUpdateEvent.Link.Attrs().Name
+					klog.Infof("Vrf manager: link update received for interface %s", ifName)
+					err = vrfm.syncLink(linkUpdateEvent.Link)
+					if err != nil {
+						klog.Errorf("Vrf manager: Error syncing link %s update event: %v", ifName, err)
+					}
+				case <-linkSyncTimer.C:
+					unsubscribe(doneCh, linkUpdateCh)
+					if err = vrfm.reconcile(); err != nil {
+						klog.Errorf("Vrf manager: Error while reconciling vrfs: %v", err)
+					}
+				case <-stopChan:
+					unsubscribe(doneCh, linkUpdateCh)
+					return nil
+				}
+			}
+		})
+		if err != nil {
+			klog.Errorf("Vrf manager: failed to run link reconcile goroutine: %v", err)
+		}
+	}()
+	klog.Info("Vrf manager is running")
+	return nil
+}
+
+func (vrfm *Controller) reconcile() error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+	start := time.Now()
+	defer func() {
+		klog.V(5).Infof("Vrf Manager: reconciling VRFs took %v", time.Since(start))
+	}()
+	vrfsToKeep := make(map[string]*vrf)
+	for _, vrf := range vrfm.vrfs {
+		if vrf.delete {
+			err := vrfm.deleteVRF(vrf)
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		err := vrfm.sync(vrf)
+		if err != nil {
+			return err
+		}
+		vrfsToKeep[vrf.name] = vrf
+	}
+	vrfm.vrfs = vrfsToKeep
+	return nil
+}
+
+func (vrfm *Controller) syncLink(link netlink.Link) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+	vrf, ok := vrfm.vrfs[link.Attrs().Name]
+	if !ok {
+		return nil
+	}
+	return vrfm.sync(vrf)
+}
+
+func (vrfm *Controller) sync(vrf *vrf) error {
+	if vrf == nil {
+		return nil
+	}
+	vrfLink, err := util.GetNetLinkOps().LinkByName(vrf.name)
+	if err == nil {
+		if vrfLink.Type() != "vrf" {
+			return errors.New("node has another non vrf device with same name")
+		}
+		vrfDev, ok := vrfLink.(*netlink.Vrf)
+		if ok && vrfDev.Table != vrf.table {
+			return errors.New("found conflict with existing vrf device table id")
+		}
+	}
+	// Create Vrf device if it doesn't exist.
+	if util.GetNetLinkOps().IsLinkNotFoundError(err) {
+		vrfLink = &netlink.Vrf{
+			LinkAttrs: netlink.LinkAttrs{Name: vrf.name},
+			Table:     vrf.table,
+		}
+		err = util.GetNetLinkOps().LinkAdd(vrfLink)
+	}
+	if err != nil {
+		return err
+	}
+	vrfLink, err = util.GetNetLinkOps().LinkByName(vrf.name)
+	if err != nil {
+		return err
+	}
+	if vrfLink.Attrs().OperState != netlink.OperUp {
+		if err = util.GetNetLinkOps().LinkSetUp(vrfLink); err != nil {
+			return err
+		}
+	}
+	// Ensure enslave interface is synced up with the cache.
+	var existingEnslave string
+	links, err := util.GetNetLinkOps().LinkList()
+	if err != nil {
+		return err
+	}
+	for _, link := range links {
```

Comment:
id put in a func to make it more readable and return back the name once you got it instead of continuing like here.

Reply:
done